### PR TITLE
Fix auto_mixed_precision for in-graph train loops

### PR DESCRIPTION
**Pull request by benbarsdell**
Models with an in-graph while loop around the training update were observed to produce incorrect results when auto_mixed_precision was enabled. This was due to Casts being inserted between (non-resource-) Variable read nodes and Enter nodes, which breaks the model behavior.

This commit fixes the issue by preventing Enter nodes from being converted to fp16 if they are fed from Variable read nodes.

Also adds a test that demonstrates the issue with an in-graph training loop.

Attn. @reedwm 
cc. @nluehr

### CODE REVIEW

File: tensorflow/python/grappler/auto_mixed_precision_test.py [hunk 1]
```diff
@@ -616,6 +621,49 @@ def test_noninlined_funcdef(self):
...
+    Enter ops that appear in front of reads from non-resource variables. See
+    the use of NodeImplicitlyReadsVariable in auto_mixed_precision.cc.
+    """
+    if test.is_gpu_available(cuda_only=True):
+      random_seed.set_random_seed(1234)
+      np.random.seed(1234)
+      num_iter, bs, nchan, nclass = 100, 64, 32, 100
+
+      data = np.random.normal(size=(bs * num_iter, nchan)).astype(np.float32)
+      labels = np.random.randint(nclass, size=(bs * 100,))
```
Comment (reedwm) [inline tensorflow/python/grappler/auto_mixed_precision_test.py]: Should this 100 be `num_iter`?

File: tensorflow/python/grappler/auto_mixed_precision_test.py [hunk 2]
```diff
@@ -616,6 +621,49 @@ def test_noninlined_funcdef(self):
...
+        loss = losses.sparse_softmax_cross_entropy(yt, y)
+        opt = adam.AdamOptimizer()
+        train_op = opt.minimize(loss)
+        with ops.control_dependencies([train_op]):
+          loss = array_ops.identity(loss)
+        return loss, i, j
+
+      begin, end = constant_op.constant(0), constant_op.constant(num_iter)
+      loss, _, _ = control_flow_ops.while_loop(
+          lambda loss, i, j: math_ops.less(i, j),
```
Comment (reedwm) [inline tensorflow/python/grappler/auto_mixed_precision_test.py]: I don't think `j` has to be a loop variable. Instead you can just use `end`.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/python/grappler/auto_mixed_precision_test.py (hunk 1 above)
Comment (benbarsdell) [reply tensorflow/python/grappler/auto_mixed_precision_test.py]: Fixed.

File: tensorflow/python/grappler/auto_mixed_precision_test.py (hunk 2 above)
Comment (benbarsdell) [reply tensorflow/python/grappler/auto_mixed_precision_test.py]: Done.

**Review by reedwm — DISMISSED** [review]

**Comment by reedwm** [comment]
It looks like the `//tensorflow/python:auto_mixed_precision_test_gpu` test for the "Linux GPU" check is failing.

**Review by reedwm — APPROVED** [review]

**Comment by benbarsdell** [comment]
"RuntimeError: Attempting to capture an EagerTensor without building a function."
I don't know how to interpret that, and I can't reproduce the failure in my build. I'm not sure if this is because I'm running the tests differently, or because my branch is out of date. Any guesses as to which is more likely?

**Comment by reedwm** [comment]
I think this issue is occurring with TF2 only. Can you try running with TF 2?

This is probably a bug somewhere in TensorFlow. The `run_deprecated_v1` decorator only turns on graph mode, but still keeps the other TF 2 behavior I think. Somewhere an EagerTensor is probably being created despite graph mode being on.

**Comment by benbarsdell** [comment]
I pulled master branch and ran with "bazel test --config=v2 ..." but I still can't reproduce this test failure. Any suggestions?

**Comment by reedwm** [comment]
I will try debugging

**Pull request merged**